### ISSUE TITLE
AbstractLogstashTcpSocketAppender ignores Encoder#headerBytes and Encoder#footerBytes

### ISSUE THREAD

**Issue by brenuart** [logstash/logstash-logback-encoder#466]
`AbstractLogstashTcpSocketAppender` ignores `Encoder#headerBytes` and `Encoder#footerBytes` when encoding log events. 
But.... when should these header/footer bytes be appended?
- should the prefix be added when opening the stream and the footer when closing it? The appender never closes the stream by itself (except when reconnecting to the master from a secondary)...
- should the be added before/after every event? In this case, what's the added value compared to including them inside the `encode` method?

I can't see any scenario where prefix/suffix would be useful in the context of the TcpSocketAppender...

**Comment by philsttr** [comment]
> should the prefix be added when opening the stream and the footer when closing it? The appender never closes the stream by itself (except when reconnecting to the master from a secondary)...

Yes.  The javadoc of the Encoder.[headerBytes](https://github.com/qos-ch/logback/blob/bbc783e6d68482abe8e7dff0d06708175e971b91/logback-core/src/main/java/ch/qos/logback/core/encoder/Encoder.java#L33-L34)/[footerBytes](https://github.com/qos-ch/logback/blob/bbc783e6d68482abe8e7dff0d06708175e971b91/logback-core/src/main/java/ch/qos/logback/core/encoder/Encoder.java#L48-L49) indicates that the return values of the methods should be added to the stream upon open/close, respectively. 

In addition to the reconnect scenario, the appender should close the stream when the appender is stopped.

> should the be added before/after every event? 

No, at least according to the javadoc of the Encoder.headerBytes/footerBytes methods.


Also note that CompositeJsonEncoder has a separate prefix/suffix concept that can be used to add data before/after each event.  Prefix/suffix are useful if you need to add non-json data around the json event.

**Comment by brenuart** [comment]
Ok. So they are header/footer of the STREAM - not the EVENT.

I still don’t see how this would fit in the context of the TCP appender.
- prefix can be output when opening a new connection - OK
- suffix can be sent when actively closing the output stream - OK. But this means the connection must be “alive” - doesn’t work when the appender is disconnected or closes the socket because of a “dead peer” (that may not be actually dead but the network is down). I mean, I don’t see how a receiver can rely on the suffix...

Anyway, I don’t see any actual use case. And I even wonder if it is not “too dangerous” to implement this support. I mean, what would happen if someone uses an Encoder that actually supports these header/footers… This may be a breaking change I’m afraid.

I’m starting to wonder why I created this issue… may be I needed to develop the problem statement to realise it is a “no sense” ;-)
